sdv/single_table/base.py

```diff
@@ -42,3 +46,175 @@ def get_parameters(self):
+
+        return errors
+
+    @staticmethod
+    def _update_invalid_values(invalid_values):
+        invalid_values = sorted(invalid_values, key=lambda x: str(x))
+        if len(invalid_values) > 3:
+            return invalid_values[:3] + [f'+ {len(invalid_values) - 3} more']
+
+        return invalid_values
+
+    def _validate_key_values_are_unique(self, data):
+        errors = []
+        keys = self._get_primary_plus_alternate_keys()
+        for key in sorted(keys):
+            repeated_values = set(data[key][data[key].duplicated()])
+            if repeated_values:
+                repeated_values = self._update_invalid_values(repeated_values)
+                errors.append(f"Key column '{key}' contains repeating values: {repeated_values}")
+
+        return errors
+
+    def _validate_context_columns(self, data):
+        errors = []
+        context_column_names = self._data_processor._model_kwargs.get('context_columns')
+        if context_column_names:
+            sequence_key_names = sorted(self._get_sequence_keys())
+            for sequence_key, group in data.groupby(sequence_key_names):
+                if len(group.groupby(context_column_names)) != 1:
+                    invalid_context_cols = {}
+                    for context_column_name in context_column_names:
+                        values = set(group[context_column_name])
+                        if len(values) != 1:
+                            invalid_context_cols[context_column_name] = values
+
+                    errors.append(
+                        f'Context column(s) {invalid_context_cols} are changing '
+                        f'inside the sequence keys ({sequence_key_names}: {sequence_key}).'
+                    )
+
+        return errors
+
+    @staticmethod
+    def _is_numerical(value):
+        return pd.isna(value) | pd.api.types.is_float(value) | pd.api.types.is_integer(value)
+
+    @staticmethod
+    def _is_boolean(value):
+        return True if pd.isna(value) | (value is True) | (value is False) else False
+
+    def _validate_column(self, column):
+        """Validate values of the column satisfy its sdtype properties."""
+        errors = []
+        sdtype = self.metadata._columns[column.name]['sdtype']
+
+        # boolean values must be True/False, None or missing values
+        # int/str are not allowed
+        if sdtype == 'boolean':
+            valid = column.apply(self._is_boolean)
+            invalid_values = set(column[~valid])
+            if invalid_values:
+                invalid_values = self._update_invalid_values(invalid_values)
+                errors.append(
+                    f"Invalid values found for boolean column '{column.name}': {invalid_values}.")
+
+        # numerical values must be int/float, None or missing values
+        # str/bool are not allowed
+        if sdtype == 'numerical':
+            valid = column.apply(self._is_numerical)
+            invalid_values = set(column[~valid])
```

Comment:
Any reason not to have the methods just return the set of invalid values themselves?

Reply:
Seemed a little easier to read to me, though I don't have a preference. I can change it if you want?

Reply:
I just notice they all follow the pattern of checking if they are the type, filtering out the ones that aren't and then raising an error with them. You could have methods that say `_get_invalid_numerical_values`, `_get_invalid_datetime_values`... It's minor though